libraries/apollo-ast/src/commonMain/kotlin/com/apollographql/apollo3/ast/introspection/meta_introspection_reader.kt

```diff
@@ -0,0 +1,156 @@
+   */
+  SchemaDescription,
+
+  /**
+   * `__Type.specifiedByURL`, introduced in [October 2021](https://spec.graphql.org/October2021/)
+   */
+  TypeSpecifiedByURL,
+
+  /**
+   * `__Directive.isRepeatable`, introduced in [October 2021](https://spec.graphql.org/October2021/)
+   */
+  DirectiveIsRepeatable,
+
+  /**
+   * `__Type.inputFields`, `includeDeprecated` argument
+   */
+  TypeInputFieldsIncludeDeprecated,
+
+  /**
+   * `__Directive.args`, `includeDeprecated` argument
+   */
+  DirectiveArgsIncludeDeprecated,
+
+  /**
+   * `__Field.args`, `includeDeprecated` argument
+   */
+  FieldArgsIncludeDeprecated,
+
+  /**
+   * `__InputValue.isDeprecated`
+   */
+  InputValueIsDeprecated,
+
+  /**
+   * `__InputValue.deprecationReason`
+   */
+  InputValueDeprecatedReason,
+
+  /**
+   * `__Type.isOneOf`, introduced in [OneOf Input Objects RFC](https://github.com/graphql/graphql-spec/pull/825/)
+   */
+  TypeIsOneOf
+}
+
+@ApolloInternal
+fun String.toIntrospectionCapabilities(): Set<IntrospectionCapability> {
+  val introspection = this.toMIntrospection()
+  return buildSet {
+    val types = introspection.data.__schema.types
+    types.firstOrNull { it.name == "__Schema" }?.let { schema ->
+      if (schema.fields.orEmpty().any { it.name == "description" }) {
+        add(SchemaDescription)
+      }
+    }
+    types.firstOrNull { it.name == "__Type" }?.let { type ->
+      if (type.fields.orEmpty().any { it.name == "specifiedByURL" }) {
+        add(TypeSpecifiedByURL)
+      }
+      if (type.fields.orEmpty().any { it.name == "isOneOf" }) {
+        add(TypeIsOneOf)
+      }
+      type.fields.orEmpty().firstOrNull{ it.name == "inputFields" }?.let { inputFields ->
+        if (inputFields.args.any { it.name == "includeDeprecated" }) {
+          add(TypeInputFieldsIncludeDeprecated)
+        }
+      }
+    }
+    types.firstOrNull { it.name == "__Directive" }?.let { directive ->
+      if (directive.fields.orEmpty().any { it.name == "isRepeatable" }) {
+        add(DirectiveIsRepeatable)
+      }
+      directive.fields.orEmpty().firstOrNull{ it.name == "args" }?.let { args ->
+        if (args.args.any { it.name == "includeDeprecated" }) {
+          add(DirectiveArgsIncludeDeprecated)
+        }
+      }
+    }
+    types.firstOrNull { it.name == "__Field" }?.let { field ->
+      field.fields.orEmpty().firstOrNull{ it.name == "args" }?.let { args ->
+        if (args.args.any { it.name == "includeDeprecated" }) {
+          add(FieldArgsIncludeDeprecated)
+        }
+      }
+    }
+    types.firstOrNull { it.name == "__InputValue" }?.let { inputValue ->
+      if (inputValue.fields.orEmpty().any { it.name == "isDeprecated" }) {
+        add(InputValueIsDeprecated)
+      }
+      if (inputValue.fields.orEmpty().any { it.name == "deprecationReason" }) {
+        add(InputValueDeprecatedReason)
+      }
+    }
+  }
+}
+
+private val json = Json {
+  // be robust to errors: [] keys
+  ignoreUnknownKeys = true
```

Comment:
What happens if a field errors (very unlikely to happen for an introspection query but 🤷 )?  Should we output an error instead?

Reply:
If you get an error here, something is very wrong, so it would make sense to fail early I think.